### PR TITLE
Add support for field directives

### PR DESCRIPTION
**Pull request by sforward**
## Summary
This PR starts where #703 was left off. It is a non-breaking implementation of the field directive, by overriding the default field resolver.

*Locations*
There are two types of directive locations: type system (or schema) directive locations and executable directive locations. The type system directives are not supported by webonyx/graphql-php, [see here](https://github.com/webonyx/graphql-php/issues/588#issuecomment-557806121). This isn't a real problem though, since similar functionality can be achieved via the types, queries, middleware, etc of this package. I have added a warning for this in the readme.

I decided to make a PR for the field location only, since the other locations will act on different parts of the query execution.

*Built-in directives*
The `include` and `skip` directives already worked, so I added these to the readme as well.

---

Type of change:
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update
- [ ] Misc. change (internal, infrastructure, maintenance, etc.)

Checklist:
- [x] Existing tests have been adapted and/or new tests have been added
- [x] Add a CHANGELOG.md entry
- [x] Update the README.md
- [x] Code style has been fixed via `composer fix-style`

### CODE REVIEW

**Comment by illambo** [comment]
Hi! it is a pleasure that someone has had the desire to pick up the thing again. 
Has nothing else been introduced compared to the old PR?

**Comment by sforward** [comment]
The differences with the previous PR are:

- Directives work with arguments
- Internal directives are still available if you have custom directives registered
- Definition of the custom directive classes are made similar to Type classes (with $attributes and args method)
- The logic needed to apply the directives is separated from the field resolver code

@illambo I hope this answers your question!

**Comment by illambo** [comment]
Yes! thanks for the quick reply and for continuing development in this.

**Comment by mfn** [comment]
Just a quick no-update update: I've not forgotten, still on my list, but it's a "bad time of the year" timing and it's not that I just have to review handful lines of code. Thanks for you patience 🙏🏼

**Review by mfn — COMMENTED** [review]
Wow, really nice PR! This is a preliminary review until I better understand it, please see my feedback below and inline!

I was able to follow _some of the parts_ but not yet all of it.

I'm surprised to understand (or: maybe I got it wrong) that the graphql-php default field resolver is responsible for resoling the field directives. I would have though this would be built-in somehow.
Because if I got this right, this would mean directive support depends on the default field resolver and if a custom one is provided and does not include the support for it, it won't work.

Can you explain this in more detail?

Also, do you have an idea what it would need to support other kind of directives?

Thank you!

---

- Please import all FQCN (I saw it multiple times but only gave feedback in one place)
- I noticed that the `FieldResolver` and `DirectiveHandler` are not registered in the service provider. This means if running under a different architecture, like in Laravel Octane via e.r. roadrunner, we create multiple instances but by and large, they never really change?
  TL;DR: my question here is: can they be made singletons?

File: src/GraphQL.php [hunk 1]
```diff
@@ -378,10 +380,23 @@ public function buildSchemaFromConfig(array $schemaConfig): Schema
...
+
+        foreach ($schemaDirectives as $name => $class) {
+            $directive = $this->app->make($class);
+
+            if (!\is_string($name)) {
+                $name = $directive->name;
+            }
+
+            $directives[$name] = $directive;
+        }
```
[the remark this answers is not in the document]
Comment (sforward) [reply src/GraphQL.php]: I've chosen this approach to be consistent with how queries and mutations can be added to the schema here. But fine by me to change is as suggested.

File: src/Support/FieldResolver/DirectiveHandler.php [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+            $this->directives[$directive->name] = $directive;
+        }
+
+        return $this;
+    }
+
+    public function getDirective(string $name): ?Directive
+    {
+        return $this->directives[$name] ?? null;
+    }
```
[the remark this answers is not in the document]
Comment (sforward) [reply src/Support/FieldResolver/DirectiveHandler.php:31]: With the designed usage you can never run into the situation that the directive isn't present, because the query validation will be executed before. But since the method is public I added it to prevent the "undefined index" exception from being thrown, and anyone can decide how to handle the null value in its application.

**Comment by sforward** [comment]
> I'm surprised to understand (or: maybe I got it wrong) that the graphql-php default field resolver is responsible for resoling the field directives. I would have though this would be built-in somehow.
> Because if I got this right, this would mean directive support depends on the default field resolver and if a custom one is provided and does not include the support for it, it won't work.

While working on this, and digging into webonyx/graphl-php, I was surprised as well. The built-in directives skip and include are hard-coded in graphql-php, by leaving out fields before they are resolved. No other directive support is available, and a related issue is open for a long time. The field resolver is the suggested approach according [their documentation](https://webonyx.github.io/graphql-php/type-definitions/directives/#custom-directives). That's the reason that the possible directive locations are limited currently. 

If someone defines a custom field resolver, the directive support will indeed no longer be available. I will add a warning for this in de readme, if you want to continue with this PR. I can understand if you see too many objections.

**Comment by mfn** [comment]
> If someone defines a custom field resolver, the directive support will indeed no longer be available.

If I understand it correctly, it would boil down to having to delegate only one call to your architecture to support them as I see in https://github.com/rebing/graphql-laravel/blob/07056b69e4b16c3ded7505c1fa63c176c9ab60b4/src/Support/FieldResolver/FieldResolver.php#L38

This would be acceptable to me although I'm not yet sure how a final solution regarding the default field resolver should work:
- as it does right now: shift responsibility to dev if they use a custom resolver, but then we should make integration of directives easier
- always wrap it as part of this framework and callback to the user-defined one, so directives are always applied / controllable via the framework

> if you want to continue with this PR. I can understand if you see too many objections.

Quite some time has passed and even I had to re-read all my feedback to get the picture again. My gut feeling is that we might need some more iteration(s) to nail it, but otherwise
I would gladly have this PR and have its support in the framework.

**Comment by mfn** [comment]
@sforward are you still interested in continuing this PR?

**Comment by sforward** [comment]
Sorry it's been a while. The linked issue at webonyx/graphql-php was closed yesterday, so I don't expect we can achieve full support for directives here. I will close the PR.
